File: zig/src/scheduler.zig
//! NexusAOS Priority Scheduler (Zig)
//! Binary heap-based priority queue with O(log n) enqueue/dequeue.

const std = @import("std");
const Allocator = std.mem.Allocator;

pub const Priority = enum(u8) {
    Low = 0,
    Normal = 1,
    High = 2,
    Critical = 3,
};

pub const SchedulerEntry = struct {
    task_id: [16]u8, // UUID bytes
    priority: Priority,
    enqueued_at: i64, // unix millis
};

pub const Scheduler = struct {
    const Self = @This();

    entries: std.ArrayList(SchedulerEntry),
    max_depth: usize,
    allocator: Allocator,

    pub fn init(allocator: Allocator, max_depth: usize) Self {
        return Self{
            .entries = std.ArrayList(SchedulerEntry).init(allocator),
            .max_depth = max_depth,
            .allocator = allocator,
        };
    }

    pub fn deinit(self: *Self) void {
        self.entries.deinit();
    }

    /// Enqueue a task. Returns error if queue is full.
    pub fn enqueue(self: *Self, task_id: [16]u8, priority: Priority) !void {
        if (self.entries.items.len >= self.max_depth) return error.QueueFull;

        const now = std.time.milliTimestamp();
        try self.entries.append(.{
            .task_id = task_id,
            .priority = priority,
            .enqueued_at = now,
        });

        // Sift up
        var idx = self.entries.items.len - 1;
        while (idx > 0) {
            const parent = (idx - 1) / 2;
            if (self.compare(idx, parent)) break;
            std.mem.swap(SchedulerEntry, &self.entries.items[idx], &self.entries.items[parent]);
            idx = parent;
        }
    }

    /// Dequeue the highest-priority task.
    pub fn dequeue(self: *Self) ?SchedulerEntry {
        if (self.entries.items.len == 0) return null;

        const result = self.entries.items[0];
        const last = self.entries.pop();

        if (self.entries.items.len > 0) {
            self.entries.items[0] = last;
            self.siftDown(0);
        }

        return result;
    }

    /// Get current queue depth.
    pub fn queueDepth(self: *const Self) usize {
        return self.entries.items.len;
    }

    /// Cancel a task by task_id.
    pub fn cancel(self: *Self, task_id: [16]u8) bool {
        for (self.entries.items, 0..) |entry, i| {
            if (std.mem.eql(u8, &entry.task_id, &task_id)) {
                _ = self.entries.swapRemove(i);
                if (i < self.entries.items.len) {
                    self.siftDown(i);
                    // Also try sifting up
                    var idx = i;
                    while (idx > 0) {
                        const parent = (idx - 1) / 2;
                        if (self.compare(idx, parent)) break;
                        std.mem.swap(SchedulerEntry, &self.entries.items[idx], &self.entries.items[parent]);
                        idx = parent;
                    }
                }
                return true;
            }
        }
        return false;
    }

    fn siftDown(self: *Self, idx: usize) void {
        var i = idx;
        while (true) {
            var largest = i;
            const left = 2 * i + 1;
            const right = 2 * i + 2;

            if (left < self.entries.items.len and self.compare(left, largest)) {
                largest = left;
            }
            if (right < self.entries.items.len and self.compare(right, largest)) {
                largest = right;
            }

            if (largest == i) break;
            std.mem.swap(SchedulerEntry, &self.entries.items[i], &self.entries.items[largest]);
            i = largest;
        }
    }

    /// Returns true if a has higher priority than b.
    fn compare(self: *const Self, a: usize, b: usize) bool {
        const ea = self.entries.items[a];
        const eb = self.entries.items[b];
        if (@intFromEnum(ea.priority) != @intFromEnum(eb.priority)) {
            return @intFromEnum(ea.priority) > @intFromEnum(eb.priority);
        }
        return ea.enqueued_at < eb.enqueued_at; // older first
    }
};

// ── C ABI ──────────────────────────────────────────

export fn scheduler_create(max_depth: usize) ?*Scheduler {
    const allocator = std.heap.c_allocator;
    const s = allocator.create(Scheduler) catch return null;
    s.* = Scheduler.init(allocator, max_depth);
    return s;
}

export fn scheduler_enqueue(s: *Scheduler, task_id: *const [16]u8, priority: u8) bool {
    s.enqueue(task_id.*, @enumFromInt(priority)) catch return false;
    return true;
}

export fn scheduler_dequeue(s: *Scheduler) ?SchedulerEntry {
    return s.dequeue();
}

export fn scheduler_queue_depth(s: *const Scheduler) usize {
    return s.queueDepth();
}

export fn scheduler_cancel(s: *Scheduler, task_id: *const [16]u8) bool {
    return s.cancel(task_id.*);
}

export fn scheduler_deinit(s: *Scheduler) void {
    s.deinit();
    std.heap.c_allocator.destroy(s);
}

File: src/runtime/scheduler.rs
use std::{
    cmp::Ordering,
    collections::BinaryHeap,
    sync::{
        Mutex,
        atomic::{AtomicUsize, Ordering as AtomicOrdering},
    },
};

use chrono::{DateTime, Utc};
use tokio_util::sync::CancellationToken;

use crate::{
    error::TaskError,
    task::{Priority, TaskId},
};

#[derive(Debug)]
pub struct SchedulerEntry {
    pub task_id: TaskId,
    pub priority: Priority,
    pub enqueued_at: DateTime<Utc>,
    pub cancellation: CancellationToken,
}

impl PartialEq for SchedulerEntry {
    fn eq(&self, other: &Self) -> bool {
        self.task_id == other.task_id
    }
}

impl Eq for SchedulerEntry {}

impl PartialOrd for SchedulerEntry {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for SchedulerEntry {
    fn cmp(&self, other: &Self) -> Ordering {
        // Higher priority first
        match self.priority.cmp(&other.priority) {
            Ordering::Equal => {
                // If priorities are equal, older tasks (smaller timestamp) have higher precedence
                other.enqueued_at.cmp(&self.enqueued_at)
            }
            other_order => other_order,
        }
    }
}

/// Priority-based task scheduler with depth limits.
pub struct Scheduler {
    queue: Mutex<BinaryHeap<SchedulerEntry>>,
    max_depth: usize,
    active_count: AtomicUsize,
}

impl Scheduler {
    pub fn new(max_depth: usize) -> Self {
        Self { queue: Mutex::new(BinaryHeap::new()), max_depth, active_count: AtomicUsize::new(0) }
    }

    /// Enqueue a task. Returns error if queue is full.
    pub async fn enqueue(
        &self,
        task_id: TaskId,
        priority: Priority,
    ) -> Result<CancellationToken, TaskError> {
        let mut queue = self.queue.lock().unwrap_or_else(|e| e.into_inner());

        if queue.len() >= self.max_depth {
            return Err(TaskError::QueueFull { max_depth: self.max_depth });
        }

        let token = CancellationToken::new();
        let entry = SchedulerEntry {
            task_id,
            priority,
            enqueued_at: Utc::now(),
            cancellation: token.clone(),
        };

        queue.push(entry);
        self.active_count.fetch_add(1, AtomicOrdering::SeqCst);

        Ok(token)
    }

    /// Dequeue the highest-priority task.
    pub async fn dequeue(&self) -> Option<SchedulerEntry> {
        let mut queue = self.queue.lock().unwrap_or_else(|e| e.into_inner());
        let entry = queue.pop()?;
        self.active_count.fetch_sub(1, AtomicOrdering::SeqCst);
        Some(entry)
    }

    /// Get current queue depth.
    pub fn queue_depth(&self) -> usize {
        self.active_count.load(AtomicOrdering::SeqCst)
    }

    /// Cancel a specific task.
    pub async fn cancel(&self, task_id: &TaskId) -> bool {
        let mut queue = self.queue.lock().unwrap_or_else(|e| e.into_inner());

        // BinaryHeap doesn't support easy removal by identity, so we rebuild it
        let mut found = false;
        let mut new_heap = BinaryHeap::new();

        for entry in queue.drain() {
            if entry.task_id == *task_id {
                entry.cancellation.cancel();
                found = true;
                self.active_count.fetch_sub(1, AtomicOrdering::SeqCst);
            } else {
                new_heap.push(entry);
            }
        }

        *queue = new_heap;
        found
    }

    /// Drain all entries (for shutdown).
    pub async fn drain(&self) -> Vec<SchedulerEntry> {
        let mut queue = self.queue.lock().unwrap_or_else(|e| e.into_inner());
        let drained: Vec<_> = queue.drain().collect();
        self.active_count.store(0, AtomicOrdering::SeqCst);
        drained
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_enqueue_dequeue() {
        let scheduler = Scheduler::new(10);
        let task_id = TaskId::new();
        let _ = scheduler.enqueue(task_id, Priority::Normal).await.unwrap();
        assert_eq!(scheduler.queue_depth(), 1);

        let entry = scheduler.dequeue().await.unwrap();
        assert_eq!(entry.task_id, task_id);
        assert_eq!(scheduler.queue_depth(), 0);
    }

    #[tokio::test]
    async fn test_overflow() {
        let scheduler = Scheduler::new(1);
        let _ = scheduler.enqueue(TaskId::new(), Priority::Normal).await.unwrap();
        let result = scheduler.enqueue(TaskId::new(), Priority::Normal).await;
        assert!(matches!(result, Err(TaskError::QueueFull { .. })));
    }

    #[tokio::test]
    async fn test_cancel() {
        let scheduler = Scheduler::new(10);
        let task_id = TaskId::new();
        let _ = scheduler.enqueue(task_id, Priority::Normal).await.unwrap();
        assert!(scheduler.cancel(&task_id).await);
        assert_eq!(scheduler.queue_depth(), 0);
        assert!(scheduler.dequeue().await.is_none());
    }

    #[tokio::test]
    async fn test_priority_ordering() {
        let scheduler = Scheduler::new(10);
        let task1 = TaskId::new();
        let task2 = TaskId::new();
        let task3 = TaskId::new();

        let _ = scheduler.enqueue(task1, Priority::Low).await.unwrap();
        let _ = scheduler.enqueue(task2, Priority::High).await.unwrap();
        let _ = scheduler.enqueue(task3, Priority::Normal).await.unwrap();

        assert_eq!(scheduler.dequeue().await.unwrap().task_id, task2);
        assert_eq!(scheduler.dequeue().await.unwrap().task_id, task3);
        assert_eq!(scheduler.dequeue().await.unwrap().task_id, task1);
    }
}
